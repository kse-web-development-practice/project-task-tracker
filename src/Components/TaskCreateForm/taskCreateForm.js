import React, { useState } from 'react'
import styles from './taskCreateForm.module.css'
import { FormInput } from '../FormInput/formInput'
import { Button } from '../Button/button'
import { FormTextarea } from '../FormTextarea/formTextarea'

export const TaskCreateForm = () => {
  const [title, setTitle] = useState('')
  const [description, setDescription] = useState('')
  const [deadline, setDeadline] = useState('')
  const [error, setError] = useState('')

  const handleSubmit = (event) => {
    event.preventDefault()
    if (!title) {
      setError('Enter task title.')
    } else {
      // post func
      setError('')
    }
  }

  return (
    <form className={styles.taskCreateForm} onSubmit={handleSubmit}>
      {error && <div className="error">{error}</div>}
      <FormInput
        labelText="Task Title:"
        type="text"
        value={title}
        onChange={(e) => setTitle(e.target.value)}
      />
      <FormTextarea
        labelText="Task Description:"
        value={description}
        onChange={(e) => setDescription(e.target.value)}
      />
      <FormInput
        labelText="Deadline:"
        type="date"
        value={deadline}
        onChange={(e) => setDeadline(e.target.value)}
      />
      <Button isMain={true} type="submit">
        Create Task
      </Button>
    </form>
  )
}

