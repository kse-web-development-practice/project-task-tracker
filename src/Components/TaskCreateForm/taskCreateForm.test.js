import React from 'react'
import { render, screen, fireEvent, act, waitFor } from '@testing-library/react'
import { TaskCreateForm } from './taskCreateForm'
import { useNavigate } from 'react-router-dom'

jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn()
}))

describe('TaskCreateForm Component', () => {
  const renderTaskCreateForm = (createFunc) =>
    render(<TaskCreateForm createFunc={createFunc} projectId="123" />)

  test('default snapshot', () => {
    const { asFragment } = renderTaskCreateForm()
    expect(asFragment()).toMatchSnapshot()
  })

  test('error snapshot', () => {
    const { asFragment } = renderTaskCreateForm()

    fireEvent.click(screen.getByText('Create Task'))

    expect(asFragment()).toMatchSnapshot()
  })

  test('valid submission', async () => {
    const createFunc = jest.fn()

    const mockNavigate = jest.fn()
    useNavigate.mockReturnValue(mockNavigate)

    renderTaskCreateForm(createFunc)
    act(() => {
      fireEvent.change(screen.getByLabelText(/Task Title:/i), { target: { value: 'New Task' } })
      fireEvent.change(screen.getByLabelText(/Task Description:/i), {
        target: { value: 'Task description' }
      })
      fireEvent.change(screen.getByLabelText(/Deadline:/i), { target: { value: '2024-12-31' } })

      fireEvent.click(screen.getByText('Create Task'))
    })
    await waitFor(() => {
      expect(createFunc).toHaveBeenCalledWith(
        'New Task',
        'Task description',
        '2024-12-31',
        'Low',
        false,
        '123'
      )
      expect(mockNavigate).toHaveBeenCalledWith('/tasks/123')
    })
  })
})
