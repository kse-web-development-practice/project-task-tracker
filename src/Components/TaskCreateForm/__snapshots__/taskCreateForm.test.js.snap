// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskCreateForm Component default snapshot 1`] = `
<DocumentFragment>
  <form
    class="taskCreateForm"
  >
    <label
      class="formLabel"
    >
      Task Title:
      <input
        class="formInput"
        type="text"
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Task Description:
      <textarea
        class="formTextarea"
      />
    </label>
    <label
      class="formLabel"
    >
      Deadline:
      <input
        class="formInput"
        type="date"
        value=""
      />
    </label>
    <button
      class="button main"
    >
      Create Task
    </button>
  </form>
</DocumentFragment>
`;

exports[`TaskCreateForm Component error snapshot 1`] = `
<DocumentFragment>
  <form
    class="taskCreateForm"
  >
    <div
      class="error"
    >
      Enter title and deadline.
    </div>
    <label
      class="formLabel"
    >
      Task Title:
      <input
        class="formInput"
        type="text"
        value=""
      />
    </label>
    <label
      class="formLabel"
    >
      Task Description:
      <textarea
        class="formTextarea"
      />
    </label>
    <label
      class="formLabel"
    >
      Deadline:
      <input
        class="formInput"
        type="date"
        value=""
      />
    </label>
    <button
      class="button main"
    >
      Create Task
    </button>
  </form>
</DocumentFragment>
`;
